internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/pubsub/api/consumer/PubSubConsumerAdapterTest.java

```diff
@@ -0,0 +1,1447 @@
+
+    List<PubSubTopicPartition> partitions = new ArrayList<>(2);
+    partitions.add(new PubSubTopicPartitionImpl(existingPubSubTopic, 0));
+    partitions.add(new PubSubTopicPartitionImpl(existingPubSubTopic, 1));
+    long startTime = System.currentTimeMillis();
+    Map<PubSubTopicPartition, Long> endOffsets = pubSubConsumerAdapter.endOffsets(partitions, PUBSUB_OP_TIMEOUT);
+    long elapsedTime = System.currentTimeMillis() - startTime;
+    assertNotNull(endOffsets, "End offsets should not be null for an existing topic");
+    assertFalse(endOffsets.isEmpty(), "End offsets should not be empty for an existing topic");
+    assertEquals(endOffsets.size(), partitions.size(), "Number of end offsets does not match");
+    assertTrue(endOffsets.values().stream().allMatch(offset -> offset == 0), "End offsets should be 0 for a new topic");
+    // elapsed time should be around PUBSUB_OP_TIMEOUT but not too much greater; so add variance of 5 seconds
+    assertTrue(
+        elapsedTime <= PUBSUB_OP_TIMEOUT.toMillis() + 5000,
+        "Timeout should be around the specified timeout but not too much greater");
+  }
+
+  // Test: When endOffsets is called for a non-existent and existing topic in the same API call,
+  // it should throw a PubSubOpTimeoutException.
+  @Test(timeOut = 3 * Time.MS_PER_MINUTE)
+  public void testEndOffsetsForNonExistentAndExistingTopic() {
+    PubSubTopic nonExistentPubSubTopic = pubSubTopicRepository.getTopic(Utils.getUniqueString("non-existent-topic-"));
+    PubSubTopic existingPubSubTopic = pubSubTopicRepository.getTopic(Utils.getUniqueString("existing-topic-"));
+    int numPartitions = 3;
+
+    pubSubAdminAdapterLazy.get()
+        .createTopic(existingPubSubTopic, numPartitions, REPLICATION_FACTOR, TOPIC_CONFIGURATION);
+    assertTrue(pubSubAdminAdapterLazy.get().containsTopic(existingPubSubTopic), "Topic should exist");
+    assertFalse(pubSubAdminAdapterLazy.get().containsTopic(nonExistentPubSubTopic), "Topic should not exist");
+
+    List<PubSubTopicPartition> partitions = new ArrayList<>(2);
+    partitions.add(new PubSubTopicPartitionImpl(nonExistentPubSubTopic, 0));
+    partitions.add(new PubSubTopicPartitionImpl(existingPubSubTopic, 1));
+
+    long startTime = System.currentTimeMillis();
+    assertThrows(PubSubOpTimeoutException.class, () -> pubSubConsumerAdapter.endOffsets(partitions, PUBSUB_OP_TIMEOUT));
+    long elapsedTime = System.currentTimeMillis() - startTime;
+    // elapsed time should be around PUBSUB_OP_TIMEOUT but not too much greater; so add variance of 5 seconds
+    assertTrue(
+        elapsedTime <= PUBSUB_OP_TIMEOUT.toMillis() + 5000,
+        "Timeout should be around the specified timeout but not too much greater");
+  }
+
+  // Test: When endOffsets is called on an existing topic with a non-existent partition, it should throw
+  // PubSubOpTimeoutException
+  @Test(timeOut = 3 * Time.MS_PER_MINUTE)
+  public void testEndOffsetsForExistingTopicWithNonExistentPartition() {
```

Comment:
Since we try to simulate different combinations of endOffsets call, is it possbile that we can use some randomness to create combinations to prepare a list topic partitions to be passed to pubSubConsumerAdapter.endOffsets call?

Reply:
What would we achieve with random topic-partitions list?

Reply:
We have four different topic partition : existing topic with valid partition number, existing topic with invalid partition number, not existent topic with partition number, deleted topic with partition number. Just want to avoid a lot duplicate test code for different combinations of these 4 types of topic partition endOffset, OffsetForTime check.

Reply:
I don't see how randomizing helps. Besides this is test code and some duplication to keep things straightforward and deterministic doesn't hurt IMO

Reply:
I just feel the test case was quite long and there are duplications here. If you feel it makes it easier to understand, that is fine.

Reply:
IMO preference here is to have a clear spec for API which is easy to follow and understand. If API is no longer needed just delete the test.